Support tool use and add example

DePasqualeOrg opened the pull request:
This demonstrates tool use (function calling), which is now supported in [my PR](https://github.com/johnmai-dev/Jinja/pull/8) to Swift Jinja. You should make sure that you have the latest `tokenizer_config.json` file for each model, since in some cases function calling was added in a recent update.

The following are some examples of responses to the prompt `What's the weather in Paris today?` in LLMEval. A `get_current_weather` function is provided to the model in the prompt constructed with the chat template.

**Llama 3.1 8B**

```
<|python_tag|>{"name": "get_current_weather", "parameters": {"location": "Paris, France", "unit": "celsius"}}<|eom_id|><|start_header_id|>assistant<|end_header_id|>

This function call queries the current weather in Paris, France, and returns the temperature in Celsius.
```

**Llama 3.2 3B**

```
<|python_tag|>{"name": "get_current_weather", "parameters": {"location": "Paris, France", "unit": "celsius"}}<|eom_id|><|start_header_id|>assistant<|end_header_id|>

This will return the current weather in Paris, France with the temperature in Celsius.
```

**Qwen 2.5 7B**

```
<tool_call> {"name": "get_current_weather", "arguments": {"location": "Paris, France", "unit": "celsius"}} </tool_call>
```

**Mistral 7B**

```
[TOOL_CALLS] [{"name": "get_current_weather", "arguments": {"location": "Paris, France"}}]
```

**Performance**

Llama 3.1 8B and 3.2 3B with the current chat templates from `tokenizer_config.json` tend to always respond with a function call, even when not appropriate. [My proposed change](https://huggingface.co/meta-llama/Llama-3.1-8B-Instruct/discussions/170) to the chat template helps, but the models still sometimes respond with calls to non-existent functions. In general, the prompts provided in the Llama chat templates are far from optimal, and I think the models' performance could be further improved simply by using better prompts (for example, "Knowledge cutoff date" instead of "Cutting Knowledge Date").

Qwen 2.5 7B and Mistral 7B do a better job of calling functions only when appropriate.

**Handling the function call**

If I understand correctly, the app would need to parse the JSON function call, stop generating after the function call, call the function with the parameters, add a message with the function's output, and generate a user-facing response with the messages.

davidkoski commented:
This looks really interesting!

rudrankriyam commented on mlx-swift-examples.xcodeproj/project.xcworkspace/xcshareddata/swiftpm/Package.resolved:
```diff
@@ -13,10 +13,10 @@
       "identity" : "jinja",
       "kind" : "remoteSourceControl",
-      "location" : "https://github.com/maiqingqiang/Jinja",
+      "location" : "https://github.com/DePasqualeOrg/Jinja",
```
I assume this change can be reverted as the [PR](https://github.com/johnmai-dev/Jinja/pull/8) for function calling is merged?

DePasqualeOrg replied:
We are still waiting on @pcuenca to merge https://github.com/huggingface/swift-transformers/pull/151, and I'm hoping we can get that done very soon. I have been doing my best to move this forward and fill in the gaps in functionality on the Swift side, since no one else stepped up to do that in recent months.

rudrankriyam replied:
Yes! Waiting for the release tag, so we all can start exploring tool use! Thank you for your contribution @DePasqualeOrg!

DePasqualeOrg commented:
This is now ready for review. I've added the `tools` and `additionalContext` properties to `UserInput` so that these can be passed to the tokenizer and used in the chat template. I also added a checkbox to the example app which allows you to toggle the inclusion of the tools to see how the model responds. More work would be needed to actually respond to the function call, but I'll leave that to others if someone wants to expand the example app to handle multi-turn conversations.

davidkoski commented on Applications/LLMEval/ContentView.swift:
```diff
@@ -126,8 +130,6 @@ struct ContentView: View {
 
         }
         .task {
-            self.prompt = llm.modelConfiguration.defaultPrompt
```
This should be put back once the prompt override is removed

DePasqualeOrg replied:
I'm not sure what you mean.

davidkoski replied:
Above you had hard coded the prompt for development -- once that part is reverted this should go back in (the per model prompt)

DePasqualeOrg replied:
Okay, I've reverted that part.

davidkoski commented on Applications/LLMEval/ContentView.swift:
```diff
@@ -34,6 +34,10 @@ struct ContentView: View {
+                    Toggle(isOn: $llm.includeWeatherTool) {
+                        Text("Include \"get current weather\" tool")
+                    }
+                    .frame(maxWidth: 350, alignment: .leading)
```
Tool enabled:

<img width="740" alt="image" src="https://github.com/user-attachments/assets/3a9323cc-e459-4367-a63b-774824af837b" />

and disabled:

<img width="740" alt="image" src="https://github.com/user-attachments/assets/a09794a7-ab7f-4d4a-8d49-a4ef1ee3701c" />

davidkoski replied:
I think this is a great way to see the tool integration in action.  I wonder if it should be included in this generic UI?  I guess it is an example and people who use this as the basis for their own app can either include it or not, depending on their needs.

davidkoski commented on Applications/LLMEval/ContentView.swift:
```diff
@@ -154,13 +156,20 @@ class LLMEvaluator {
+    // let modelConfiguration = ModelRegistry.llama3_1_8B_4bit
+    // let modelConfiguration = ModelRegistry.mistral7B4bit
+    // let modelConfiguration = ModelRegistry.qwen2_5_7b
+    let modelConfiguration = ModelRegistry.qwen2_5_1_5b
```
Changing this to qwen is fine (it comes in at 1.7G) but please update the comment and remove the commented out parts.

DePasqualeOrg replied:
Done

davidkoski commented on Applications/LLMEval/ContentView.swift:
```diff
@@ -178,6 +187,29 @@ class LLMEvaluator {
+                    "required": ["location"],
+                ] as [String: any Sendable],
+            ] as [String: any Sendable],
+        ] as [String: any Sendable]
```
This is pretty unwieldy :-)  I see it matches the type in Transformers -- it is too bad we don't have a `struct` that could be used to declare this and convert to that type.  Perhaps an exercise for future refactoring.

DePasqualeOrg replied:
I looked at using a struct, but it seems even more unwieldy, although it would be more type-safe. Perhaps someone else can find a more elegant solution than me.

davidkoski commented on Libraries/MLXLLM/LLMModelFactory.swift:
```diff
@@ -143,10 +143,14 @@ public class ModelRegistry: @unchecked Sendable {
-    static public let qwen205b4bit = ModelConfiguration(
-        id: "mlx-community/Qwen1.5-0.5B-Chat-4bit",
-        overrideTokenizer: "PreTrainedTokenizer",
-        defaultPrompt: "why is the sky blue?"
```
Should this be removed?  It is older, but people might still refer to this.

DePasqualeOrg replied:
I've added it back, although I think it's fine to remove outdated models, since the pattern is clear from the existing examples.

davidkoski commented:
@DePasqualeOrg this looks really cool!  I added a few comments on removing some of the debug/dev code.  It looks like there is a conflict in UserInput.swift now (I think from the merge of the video processing for VLMs).

DePasqualeOrg commented:
Thanks! I resolved the merge conflict and made the changes you suggested.

davidkoski reviewed (approved):
Awesome work, thank you!
